actionview/lib/action_view/helpers/form_with_helper.rb

```diff
@@ -0,0 +1,196 @@
+          end
+          select(method, choices, options)
+        end
+
+        def option_tags_for_select(choices, blank: false)
+          result = (blank ? tag.option("", value: "") : "").html_safe
+          result += choices.map do |choice|
+            if choice.is_a? Array
+              tag.option choice[0], value: choice[1]
+            else
+              tag.option choice, value: choice
+            end
+          end.join("\n").html_safe
+          result
+        end
+
+        private
+          def checked?(on_value, value)
+            case value
+            when TrueClass, FalseClass
+              value == !!on_value
+            when NilClass
+              false
+            when String
+              value == on_value
+            else
+              if value.respond_to?(:include?)
+                value.include?(on_value)
+              else
+                value.to_i == on_value.to_i
+              end
+            end
+          end
+
+          def translate_with_human_attribute_name(method)
+            model && model.class.respond_to?(:human_attribute_name) ? model.class.human_attribute_name(method) : nil
+          end
+
+          def submit_default_value
+            if scope
+              key = model ? (model.persisted? ? :update : :create) : :submit
+              model_name = model.respond_to?(:model_name) ? model.model_name.human : scope.to_s.humanize
+              defaults = [:"helpers.submit.#{scope}.#{key}", :"helpers.submit.#{key}", :"#{key.to_s.humanize} #{model_name}"]
+              I18n.t(defaults.shift, model: model_name, default: defaults)
+            else
+              I18n.t("helpers.submit.no_model")
+            end
+          end
+
+          def name_for(method, options)
+            scope = options.delete(:scope) { @scope }
+            index = options.delete(:index)
+            method = method.to_s.chomp("?")
+            name = scope.to_s
+            name += index ? "[#{index}]" : ""
+            name += scope.nil? ? method : "[#{method}]"
+            name += options.delete(:multiple) ? "[]" : "" #exception => file_field
+          end
+
+          def id_for(method, scope, **options)
+            scope.nil? ? method : "#{scope}_#{method}"
+          end
+
+          def options_for(method, options)
+            options.reverse_merge!(name: name_for(method, options))
+          end
+
+          def options_for_field(selector, method, args, options)
+            type = selector.split("_").first
+            options = options_for(method, options)
+            options.merge!(value: model.send(method)) if model && type != 'password'
+            options.merge!(value: args[0]) if args.size > 0
+            options[:size] = options[:maxlength] unless options.key?(:size)
+            if placeholder = placeholder(options.delete(:placeholder), method)
+              options.merge!(placeholder: placeholder)
+            end
+            options.reverse_merge!(type: type)
+          end
```

Comment:
How about this? 😁

```ruby
def #{selector}(method, *args, **options)
  tag.input options_for_field('#{selector.split("_").first}', method, args, options)
end 

def field_options(field_type, attribute, value = nil, *args, placeholder: nil, **options)
  options = default_options(attribute, options)

  if value
    options[:value] ||= value
  elsif @model && field_type.match?('password')
    options[:value] ||= @model.public_send(attribute)
  end

  options[:placeholder] = placeholder(placeholder, attribute)
  options[:size] ||= options[:max_length]
  options[:type] ||= field_type
  options
end
```

This pulls up the selector splitting, because we only need to do that once when we build the method.

Avoids `merge` calls because that allocates extra hashes and just assigning the key is clearer.